src/grt/src/fastroute/src/maze.cpp

```diff
@@ -1327,6 +1307,116 @@ float getCost(const int i,
   return cost;
 }
 
+int FastRouteCore::splitEdge(std::vector<TreeEdge>& treeedges,
+                             std::vector<TreeNode>& treenodes,
+                             int n1,
+                             int n2,
+                             int edge_n1n2)
+{
+  const int n2x = treenodes[n2].x;
+  const int n2y = treenodes[n2].y;
+
+  // create new node
+  const int new_node_id = treenodes.size();
+  TreeNode new_node;
+  new_node.x = n2x;
+  new_node.y = n2y;
+  new_node.stackAlias = treenodes[n2].stackAlias;
+
+  // create new edge
+  const int new_edge_id = treeedges.size();
+  TreeEdge new_edge;
+
+  // find one neighbor node different to n1
+  int B1;
+  int edge_n2B1;
+  if (treenodes[n2].nbr[0] == n1) {
+    B1 = treenodes[n2].nbr[1];
+    edge_n2B1 = treenodes[n2].edge[1];
+  } else if (treenodes[n2].nbr[1] == n1) {
+    B1 = treenodes[n2].nbr[0];
+    edge_n2B1 = treenodes[n2].edge[0];
+  } else {
+    B1 = treenodes[n2].nbr[0];
+    edge_n2B1 = treenodes[n2].edge[0];
+  }
+
+  // update n2 neighbor
+  int cnt = 0;
+  for (int i = 0; i < treenodes[n2].nbr_count; i++) {
+    if (treenodes[n2].nbr[i] == n1) {
+      continue;
+    }
+    if (treenodes[n2].nbr[i] == B1) {
+      treenodes[n2].nbr[cnt] = new_node_id;
+      treenodes[n2].edge[cnt] = new_edge_id;
+      cnt++;
+    } else {
+      treenodes[n2].nbr[cnt] = treenodes[n2].nbr[i];
+      treenodes[n2].edge[cnt] = treenodes[n2].edge[i];
+      cnt++;
+    }
+  }
+  treenodes[n2].nbr_count = cnt;
+
+  // change edge neighbor
+  if (treeedges[edge_n2B1].n1 == n2) {
+    treeedges[edge_n2B1].n1 = new_node_id;
+    treeedges[edge_n2B1].n1a = new_node.stackAlias;
+  } else {
+    treeedges[edge_n2B1].n2 = new_node_id;
+    treeedges[edge_n2B1].n2a = new_node.stackAlias;
+  }
+
+  // change current edge
+  if (treeedges[edge_n1n2].n1 == n2) {
+    treeedges[edge_n1n2].n1 = new_node_id;
+    treeedges[edge_n1n2].n1a = new_node.stackAlias;
+  } else {
+    treeedges[edge_n1n2].n2 = new_node_id;
+    treeedges[edge_n1n2].n2a = new_node.stackAlias;
+  }
+
+  // change node neighbor
+  for (int i = 0; i < treenodes[B1].nbr_count; i++) {
+    if (treenodes[B1].nbr[i] == n2) {
+      treenodes[B1].nbr[i] = new_node_id;
+    }
+  }
+
+  // change n1 node
+  for (int i = 0; i < treenodes[n1].nbr_count; i++) {
+    if (treenodes[n1].nbr[i] == n2) {
+      treenodes[n1].nbr[i] = new_node_id;
+    }
+  }
+
+  // config new edge
+  new_edge.len = 0;
+  new_edge.n1 = new_node_id;
+  new_edge.n1a = new_node.stackAlias;
+  new_edge.n2 = n2;
+  new_edge.n2a = treenodes[n2].stackAlias;
+  new_edge.route.type = RouteType::MazeRoute;
+  new_edge.route.routelen = 0;
+  new_edge.route.gridsX.push_back(n2x);
+  new_edge.route.gridsY.push_back(n2y);
+
+  // config new node
+  new_node.nbr_count = 3;
+  new_node.nbr[0] = B1;
+  new_node.nbr[1] = n2;
+  new_node.nbr[2] = n1;
+  new_node.edge[0] = edge_n2B1;
+  new_node.edge[1] = new_edge_id;
+  new_node.edge[2] = edge_n1n2;
```

Comment:
Will the new node you're creating always have three neighbors and three edges?

Reply:
Yes, it will always have 3 neighboring nodes because I will place it in the middle of node n2 and one of its neighbors. Then it will have as neighbors the node n,2, the node which connects that edge n1 and the neighbor node to n2.
Node n2 will always have a neighbor apart from n1 (neighbors > 1), since if it did not have one then its position does not change and it would not be used in this function.